source/developer_manual/design/processes.rst

```diff
@@ -0,0 +1,146 @@
+(``login_process_per_connection=no``), in which case the security of the
+design suffers greatly.
+
+The login processes handle SSL/TLS connections themselves completely.
+They keep proxying the connection to mail processes for the entire
+lifetime of the connection. This way if a security hole is found from
+the SSL library, an authenticated user still can't execute code outside
+the login process.
+
+See :ref:`login_processes` for more information about different settings related to login
+processes.
+
+Authentication process
+----------------------
+
+The authentication process handles everything related to the actual
+authentication: SASL authentication mechanisms, looking up and verifying
+the passwords and looking up user information.
+
+It listens for two different kinds of connections: untrusted
+authentication client connections (from login processes) and master
+connections (from master process, but also from Dovecot LDA). The client
+connections are only allowed to try to authenticate. The master
+connections are allowed to ask if an authentication request with a given
+ID was successful, and also to look up user information based on a
+username. This user lookup feature is used by Dovecot LDA.
+
+Each client connection tells their process ID to the authentication
+process in a handshake. If a connection with the same PID already
+exists, an error is logged and the new connection is refused. Although
+this makes DoS attacks possible, it won't go unnoticed for long and I
+don't see this as a real issue for now.
+
+Having the authentication process know the PID of the client connection
+allows all authentication requests to be mapped to one specific client
+connection. Since the master process knows the login process's real PID,
+it's used when asking from authentication process if the request was
+successful. This makes it impossible for a login process to try to fake
+another login process's login requests. Faking PIDs will also be quite
+pointless.
+
+Once the master process has done the verification request for a
+successful authentication request, the request is freed from memory. The
+requests are also freed about 2 minutes after their creation, regardless
+of the state they currently are in.
+
+For blocking password and user database backends (eg. MySQL) separate
+"worker processes" are used. Initially only one of them exists, but more
+are created as needed.
+:ref:`PAM <authentication-pam>`
+can be configured to use worker processes instead of doing the forking
+itself, but this isn't currently done by default and there may be
+problems related to it. Also
+:ref:`checkpassword <authentication-checkpassword>`
+currently does the forking itself.
+
+Mail processes
+--------------
+
+These processes handle the actual post-login mail handling using the
+privileges of the logged in user. It's possible to chroot these
+processes, but practically it's usually more trouble than worth.
```

Comment:
This page needs thorough updating. Some services like anvil are not mentioned at all.